File: USACO Training/3.3/Camelot - JAVA/src/camelot.java
/*
ID: cormac.1
LANG: JAVA
TASK: camelot
 */

import java.io.*;
import java.util.*;
import java.lang.*;

public class camelot {
    public static int kingDist(int px, int py, int ex, int ey) {
        return Math.min(Math.abs(px - ex), Math.abs(py - ey)) + Math.abs(Math.abs(px - ex) - Math.abs(py - ey));

    }
    static class Node implements Comparator<Node> {
        boolean hasKing;
        int x;
        int y;
        int dist;

        Node(boolean hasKing, int x, int y, int dist) {
            this.hasKing = hasKing;
            this.x = x;
            this.y = y;
            this.dist = dist;
        }

        @Override
        public int compare(Node n1, Node n2) {
            return n1.dist - n2.dist;
        }
    }

    static int R;
    static int C;

    static int KX = -1;
    static int KY = -1;

    static int[][] totalKnightCost;
    static int[][] kingCost;

    public static void main(String[] args) throws IOException {
        BufferedReader fin = new BufferedReader(new FileReader("camelot.in"));
        PrintWriter fout = new PrintWriter(new BufferedWriter(new FileWriter("camelot.out")));

        StringTokenizer s = new StringTokenizer(fin.readLine());
        R = Integer.parseInt(s.nextToken());
        C = Integer.parseInt(s.nextToken());

        String line;
        while ((line = fin.readLine()) != null) {
            s = new StringTokenizer(line);
            while (s.hasMoreTokens()) {
                int x = s.nextToken().charAt(0) - 'A';
                int y = Integer.parseInt(s.nextToken()) - 1;

                if (KX == -1) {
                    KX = x;
                    KY = y;
                    totalKnightCost = new int[R][C];
                    kingCost = new int[R][C];


                    for (int a = 0; a < C; ++a) {
                        for (int b = 0; b < R; ++b) {
                            kingCost[b][a] = kingDist(a, b, KX, KY);
                        }
                    }
                } else {
                    dijkstra(x, y);
                }
            }
        }



        int minCost = Integer.MAX_VALUE;
        for (int x = 0; x < C; ++x) {
            for (int y = 0; y < R; ++y) {
                minCost = Math.min(minCost, totalKnightCost[y][x] + kingCost[y][x]);
            }
        }

        fout.println(minCost);
        fout.close();
    }

    static int[][] KNIGHT_DELTAS = {{-2, 1}, {-1, 2}, {1, 2}, {2, 1}, {2, -1}, {1, -2}, {-1, -2}, {-2, -1}};

    public static void dijkstra(int startx, int starty) {
        int[][][] distance = new int[2][R][C];
        for (int[][] r1 : distance) {
            for (int[] r2 : r1) {
                Arrays.fill(r2, Integer.MAX_VALUE);
            }
        }
        boolean[][][] processed = new boolean[2][R][C];

        distance[0][starty][startx] = 0;

        PriorityQueue<Node> queue = new PriorityQueue<>(R * C, new Node(false, 0, 0, 0));
        queue.add(new Node(false, startx, starty, 0));

        while (queue.size() > 0) {
            Node item = queue.poll();

            int kint = item.hasKing ? 1 : 0;
            int x = item.x;
            int y = item.y;

            if (processed[kint][y][x]) continue;
            processed[kint][y][x] = true;

            // Knight moves
            for (int[] delta : KNIGHT_DELTAS) {
                int mx = delta[0];
                int my = delta[1];
                if (!(0 <= x + mx && x + mx < C && 0 <= y + my && y + my < R)) continue;

                if (distance[kint][y][x]  + 1 < distance[kint][y + my][x + mx]) {
                    distance[kint][y + my][x + mx] = distance[kint][y][x] + 1;
                    queue.add(new Node(kint == 1, x + mx, y + my,
                            distance[kint][y + my][x + mx]));
                }
            }

            // Pick up king move
            if (kint == 0) {
                if (distance[0][y][x] + kingDist(x, y, KX, KY) < distance[1][y][x]) {
                    distance[1][y][x] = distance[0][y][x] + kingDist(x, y, KX, KY);
                    queue.add(new Node(true, x, y, distance[1][y][x]));
                }
            }
        }

        for (int x = 0; x < C; ++x) {
            for (int y = 0; y < R; ++y) {
                if (totalKnightCost[y][x] != Integer.MAX_VALUE)
                    totalKnightCost[y][x] += distance[0][y][x];

                kingCost[y][x] = Math.min(kingCost[y][x], distance[1][y][x] - distance[0][y][x]);
            }
        }
    }
}

File: USACO Training/3.3/Camelot/camelot.py
"""
ID: cormac.1
LANG: PYTHON3
TASK: camelot
"""
from heapq import *

def kingDist(px, py, ex, ey):
    return min(abs(px - ex), abs(py - ey)) + abs(abs(px - ex) - abs(py - ey))

KNIGHT_DELTAS = {(-2, 1), (-1, 2), (1, 2), (2, 1), (2, -1), (1, -2), (-1, -2), (-2, -1)}

def dijkstra(point):
    # (hasKing, px, py)
    
    heap = []

    processed = [[[False
                   for i in range(C)]
                   for j in range(R)] for b in range(2)]
    
    distances = [[[float('inf')
                  for i in range(C)]
                  for j in range(R)] for b in range(2)]

    distances[False][point[1]][point[0]] = 0
    
    heappush(heap, (0, (False, point[0], point[1])))

    while heap:
        _, (hasKing, px, py) = heappop(heap)
        if processed[hasKing][py][px]: continue
        processed[hasKing][py][px] = True

        for MX, MY in KNIGHT_DELTAS:
            if not (0 <= px + MX < C and 0 <= py + MY < R): continue
            if distances[hasKing][py][px] + 1 < distances[hasKing][py + MY][px + MX]:
                distances[hasKing][py + MY][px + MX] = distances[hasKing][py][px] + 1
                heappush(heap, (distances[hasKing][py + MY][px + MX], (hasKing, px + MX, py + MY)))
        if hasKing == False:
            if distances[False][py][px] + kingDist(px, py, KX, KY) < distances[True][py][px]:
                distances[True][py][px] = distances[False][py][px] + kingDist(px, py, KX, KY)
                heappush(heap, (distances[True][py][px], (True, px, py)))
    print("DIJKSTRA")
    for row in distances[False]:
        print(row)
    print("---")
    for row in distances[True]:
        print(row)
    for y in range(R):
        for x in range(C):
            totalKnightCost[y][x] += distances[False][y][x]
            kingCost[y][x] = min(kingCost[y][x], distances[True][y][x] - distances[False][y][x])

with open("camelot.in") as file:
    R, C = map(int, file.readline().split())

    
    
    placements = file.read().split()

    KY = int(placements[1]) - 1
    KX = ord(placements[0]) - ord('A')
    
    totalKnightCost = [[0 for i in range(C)] for i in range(R)]
    kingCost        = [[kingDist(KX, KY, i, j) for i in range(C)] for j in range(R)]
    
    
    for i in range(1,len(placements) // 2):
        pointx, pointy = placements[2*i:2*i+2]
        pointx = ord(pointx) - ord('A')
        pointy = int(pointy) - 1
        dijkstra((pointx, pointy))

minMoves = float('inf')
for x in range(C):
    for y in range(R):
        minMoves = min(totalKnightCost[y][x] + kingCost[y][x], minMoves)

with open("camelot.out", "w") as file:
    file.write(str(minMoves) + "\n")
        
